### ISSUE TITLE
Convenience Methods

### ISSUE THREAD

**Issue by phortx** [vuex-orm/vuex-orm#60]
Many ORMs provide model methods like `.find()`, `.create()` or `.save()`. With Vuex-ORM we have to use rather long lines like `this.$store.dispatch('entities/users/create', { data: { ... } })` or `this.$store.getters['entities/users/query']().find()`.

What do you think about convenience methods like `.find()`, `.first()`, `.create()` and so on on the model class, that dispatch those events against vuex?

I think that would be a more comfortable way to work with vuex-orm :)

**Comment by kiaking** [comment]
It could be a nice idea. In fact, I also thought same when I first created Vuex ORM. I'm not sure if we should implement this feature, but I'm very open to discussing it.

By the way, if you just want to shorten the line, you could consider using [map helpers](https://vuex.vuejs.org/en/modules.html) (maybe you have already had though but). This is the pure Vuex way to go about shortening the method call.

But again, Vuex ORM defines Class which Vuex doesn't, so I don't think those helper methods is going to be a reason for not considering your proposal.

OK, there was two reason why I didn't implement such API (methods) to the model.

### Not sure how to (or if we should) hold Vuex Store instance

If we want to dispatch or commit Vuex inside a model, we must have access to the Store instance. So the model implementation would look something like below.

```js
class Model {
  /**
   * Get Vuex Store instance somehow.
   */
  store () {
    // ...
  }

  /**
   * Dispatch create method
   */
  create (data) {
    this.store().dispatch('entities/users/create', { data })
  }
}
```

Currently, Vuex ORM does not hold store instance in anywhere. Maybe we can inject store instance to each models being registered at the time of Vuex ORM installation, and it might solve this issue.

### I wanted users to know that it is just an ordinally Vuex module

Vuex ORM does some heavy lifting on handling data, but it is still just a Vuex Store. There are predefined actions and getters, but those are still pure Vuex actions and getters.

If you have to call a method on the model instead of dispatching action from a component, then I thought people might feel like there is special Vuex ORM magic going on. And people might think Vuex ORM has special hacky syntax.


That's being said; it has been a while since I first created this library and I think we could add your proposed methods to the model. But I want to be careful. It is going to double the maintenance cost as well.

So, what do you think about model method syntax VS map helpers syntax?

**Comment by phortx** [comment]
> Currently, Vuex ORM does not hold store instance in anywhere. Maybe we can inject store instance to each models being registered at the time of Vuex ORM installation, and it might solve this issue.

Maybe via store Plugin with param? So the developer can register the store via

```javascript
import Vue from 'vue';
import Vuex from 'vuex';
Vue.use(Vuex);

import store from 'data/store';
import VuexORM from '@vuex-orm/core';
import VuexStore from '@vuex-orm/store-plugin';
VuexORM.use(VuexStore, store)
```

I've seen similar patterns for Vuex.

> If you have to call a method on the model instead of dispatching action from a component, then I thought people might feel like there is special Vuex ORM magic going on. And people might think Vuex ORM has special hacky syntax.

This is a good reason, however, I think this can be solved by documentation, which can describe what those methods actually do: nothing except dispatching.

> So, what do you think about model method syntax VS map helpers syntax?

I always try to keep my components as clean as possible. Because they grow automatically. I'm not a big fan of boilerplate code and want to reduce it as much as possible, because boilerplate code decreases readability and maintainability. Therefore I wouldn't use the map helpers, but would be glad if there are model methods. However maybe vuex-orm could present ist like this:

> This (with normal vuex dispatching) is the way it works, but if you want (opt in) there are also model methods, which you can use to reduce boilerplate. These methods do nothing special but dispatching actions in Vuex, it's just fewer code. blabla and so on

:)

**Comment by kiaking** [comment]
> Maybe via store Plugin with param? So the developer can register the store via

Nah, I meant we are getting store instance during installation, but we must hold that store instance at somewhere. Maybe we can assign to the model, or maybe I can assign them to the database instance. Anyway, this is something I need try and figure out.

> Therefore I wouldn't use the map helpers, but would be glad if there are model methods. However maybe vuex-orm could present ist like this:

Good point. If we can solve prier issue (where to hold store instance), then it will be pretty easy to implement this. I'll try looking into this more!

**Comment by phortx** [comment]
> maybe I can assign them to the database instance

this sound very logical to me :)

Thanks!

**Comment by kiaking** [comment]
Thank you for the ideas! I think mostly good, but if we're going to implement model methods, it has to be mapped with Vuex Module method, that is;

1. create
2. insert
3. all
4. find
5. query
6. update
7. delete
8. deleteAll

I'm fine with the same method having a different name, for instance, we could add `update` and `save` that both do same thing (and a user can choose whichever they like). We already have `get`, `all` and `find`, `first` that does an exactly same thing.

Also, I think methods such as  `static fetch(filter)` or `persist()` should be defined in the plugin. Because Vuex ORM itself doesn't have "persistence layer".

Do you have any particular reason why you want to name for instance  `save(data)` instead of `update`? （again I'm fine with having two methods `save` and `update,` but I don't want to it be `save` only because the Vuex action is called `update` already.

**Comment by phortx** [comment]
> Also, I think methods such as static fetch(filter) or persist() should be defined in the plugin. Because Vuex ORM itself doesn't have "persistence layer".

I understand that argument, however I think at least providing them in vuex-orm without any implementation would define some kind of consistent API so that all plugins use the same method names and the user can swap them. If we find another way to keep this kind of consistency between the plugins, via documentation for example, that's ok for me.

> Do you have any particular reason why you want to name for instance save(data) instead of update? 

Personal taste. Coming from Rails, where models have a `save()` method, so it's just natural for me, I don't care if you want to use another name. I also like the idea of having multiple methods doing the same thing. It's the principle of least surprise from ruby.

**Comment by kiaking** [comment]
> If we find another way to keep this kind of consistency between the plugins, via documentation for example, that's ok for me.

Yeah, it could be nice if plugin author can follow such convention. However, maybe having such method already defined in Vuex ORM could oppositely affect plugin author.

That is, basically when you define a plugin, you might create a new method, or override existing method. For example, your library vuex-orm-apollo defines `fetch` which isn't present in Vuex ORM right now. There's no problem with that. Users can use that `fetch` method to interact with apollo.

But if we have `fetch` defined already in Vuex ORM, then vuex-orm-apollo will override the `fetch` method, which means users will lose their way to call `fetch` method they use to call.

In that sense, maybe we should leave that method empty for the plugin authors so that they can define an extra method that is free to use.

> Personal taste. Coming from Rails, where models have a save() method, so it's just natural for me, I don't care if you want to use another name.

Fair enough. I'm a Laravel guy too so `user.save()` would feel natural for me too. I just wanted to confirm if there was any reason other than the feelings.

**Comment by phortx** [comment]
> In that sense, maybe we should leave that method empty for the plugin authors so that they can define an extra method that is free to use.

That's exactly what I thought of :) This allows to keep the `fetch()` calls even if there is no persistence plugin  like in the test environment where you can mock it and so on.

**Comment by kiaking** [comment]
Oh I see. Yeah that's good.

**Comment by phortx** [comment]
There's still something that bothers me: In my list above all methods do either read/write from/to the store or are for syncing the store with the backend except for `destroy()` and `destroyAll()`, which do both: deleting from the store AND from the persistence. This is some kind of inconsistence and may irritate the user. Also I can imagine use cases where you want to delete a model from the store, but not on the server for whatever reasons.

So when I implement those operations in the persistence plugin, I'm forced to overwrite the implementation of both methods in the model (which is a bad practice IMHO) or I need some kind of hook to add my logic into.

That's why I'm thinking a out splitting these methods like this (very unsure about the naming here):

- `delete()`: -> `store.dispatch('entities/entity/delete', id)`
- `destroy()`: Deletes the entity form the persistence
- `static deleteAll()`: -> `store.dispatch('entities/entity/deleteAll')`
- `static destroyAll()`: Deletes all entities form the persistence

My reasoning about the naming is something like this: `delete()` (or `remove()`) is just about removing something from where it currently is (in the store) while destruction is something more "violent" and thus `destroy()` also eliminates the entity completely.

I think this is a more consequent, consistent and practicable API.

What do you think?

I'm having a model base class in my project which already provides these model methods and following the API we're defining in this ticket. The goal is to remove it, when this is implemented in Vuex-ORM itself without the necessity to change the components. When this model class is completely implemented, I will share it with you via gist maybe it helps you to implement that in Vuex-ORM :)

**Comment by kiaking** [comment]
I like the idea of using the word delete and destroy. It fits at least for me. So I guess all CRUD method has 2 each which one is save only to Vuex and the other is persist to the backend. I found few missing maybe, below is the my understanding.

| | Vuex Only | Persist |
| --- | --- | --- |
| READ | find | fetch |
| CREATE | create | persist |
| UPDATE | update | persist |
| DELETE | delete | destroy |

We have two persist though. Would it be possible? Maybe "new" instead of persist for the CREATE?

**Comment by phortx** [comment]
Everything correct :)

The point with the two persist methods is true. This is no problem if we know that this is a new record (not persisted) or not. If we don't keep that information, I'd suggest following pairs:

- create/insert and persist
- update/save and push

**Comment by kiaking** [comment]
Ah "push" is nice. 👍  Yeah now it becoming pretty solid.

**Comment by phortx** [comment]
That's probably not the final version, but I as promised, I want to share my base model class with my convenience methods, which currently work great for me.

So we have something to talk about and maybe already some code to copy'n'paste :)

https://gist.github.com/phortx/1a6d24c0e379e1490a285fa7f985d187

**Comment by kiaking** [comment]
Yeah looking good! One thing is that probably the instance methods is going to be prefixed by `$`. It is just a convention model currently has (same with Vue instance methods).

**Comment by phortx** [comment]
That's ok, I think. But just out of curiosity: what's the reason for the `$` convention?

**Comment by kiaking** [comment]
It just makes things bit safer for a user not having to worry about the conflict when defining their fields in the model.

For example, if we have `update` instance method in the model and user defines `update` fields, then the `update` method will be override and user may not be able to use that method.

And it's kinda Vue world convention. Vue also uses `$` for the instance method so I'm just following that.

**Comment by richardkmichael** [comment]
As a newcomer considering using the library, I read this issue and wanted to comment:

I found `push()` surprising as a persistence method.  `push()` feels local (Vuex/ORM only), because it is reminiscent of Array's `push()`.

I'd suggest `patch()` for the persistence "update" method because it will be a suggestion (just like `fetch()`) that the action might be remote (i.e., a reminder that you might using an HTTP endpoint).  "Patch" is also consistent with the notion of "updating".

**Comment by phortx** [comment]
Sound pretty reasonable, @richardkmichael I'd agree with you :)

**Comment by kiaking** [comment]
Finally, this is implemented at Vuex ORM 0.25.0 🎉 
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.25.0

There's no `save` method yet but I'm willing to implement it. I'll raise new issue for this! Thank you all for your patient!
